cmd/maya-exporter/app/collector/collector.go

```diff
@@ -0,0 +1,155 @@
+func (c *collector) countersList() []prometheus.Collector {
+	return []prometheus.Collector{
+		c.volumeUpTime,
+		c.connectionErrorCounter,
+		c.connectionRetryCounter,
+		c.totalReplicaCounter,
+		c.degradedReplicaCounter,
+		c.healthyReplicaCounter,
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once
+// the last descriptor has been sent. The sent descriptors fulfill the
+// consistency and uniqueness requirements described in the Desc
+// documentation. (It is valid if one and the same Collector sends
+// duplicate descriptors. Those duplicates are simply ignored. However,
+// two different Collectors must not send duplicate descriptors.) This
+// method idempotently sends the same descriptors throughout the
+// lifetime of the Collector. If a Collector encounters an error while
+// executing this method, it must send an invalid descriptor (created
+// with NewInvalidDesc) to signal the error to the registry.
+
+// Describe implements Describe method of prometheus.Collector interface.
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	for _, gauge := range c.gaugesList() {
+		gauge.Describe(ch)
+	}
+
+	for _, counter := range c.countersList() {
+		counter.Describe(ch)
+	}
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent. The
+// descriptor of each sent metric is one of those returned by
+// Describe. Returned metrics that share the same descriptor must differ
+// in their variable label values. This method may be called
+// way. Bloc	king occurs at the expense of total performance of rendering
+// concurrently and must therefore be implemented in a concurrency safe
+// all registered metrics. Ideally, Collector implementations support
+// concurrent readers.
+
+// Collect implements Collect method of prometheus.Collector interface.
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	var (
+		err         error
+		volumeStats v1.VolumeStats
+		stats       stats
+	)
+	if volumeStats, err = c.getter(); err != nil {
+		glog.Errorln(err)
+		c.setError(err)
+	}
+
+	stats = c.parser(volumeStats)
+
+	c.set(stats)
+
+	// collect the metrics extracted by collect method
+	for _, gauge := range c.gaugesList() {
+		gauge.Collect(ch)
+	}
+	for _, counter := range c.countersList() {
+		counter.Collect(ch)
+	}
+}
+
+func (c *collector) setError(err error) {
+	c.connectionErrorCounter.Inc()
+	if _, ok := err.(*connErr); ok {
+		c.connectionRetryCounter.Inc()
+	}
+}
+
+// set is used to set the values gathered from cas to
+// prometheus gauges and counters.
+func (c *collector) set(volStats stats) {
+	//	var replicaAddress, replicaMode strings.Builder
+
+	if !volStats.got {
+		glog.Warningf("%s", "setting up empty values")
```

Comment:
Can we return here by providing an `log.Info "received empty stats: did not set collector"`

Reply:
Actually, here we are setting default values in case of empty stats (i.e, 0 )